diff --git a/dlls/ntdll/signal_x86_64.c b/dlls/ntdll/signal_x86_64.c
index 7320aea..fe80a46 100644
--- a/dlls/ntdll/signal_x86_64.c
+++ b/dlls/ntdll/signal_x86_64.c
@@ -100,7 +100,7 @@ WINE_DECLARE_DEBUG_CHANNEL(relay);
  */
 static NTSTATUS virtual_unwind( ULONG type, DISPATCHER_CONTEXT *dispatch, CONTEXT *context )
 {
-    LDR_DATA_TABLE_ENTRY *module;
+    LDR_DATA_TABLE_ENTRY *module = NULL;
     NTSTATUS status;
 
     dispatch->ImageBase = 0;
diff --git a/dlls/ntdll/unwind.c b/dlls/ntdll/unwind.c
index c11bd9d..c2c950f 100644
--- a/dlls/ntdll/unwind.c
+++ b/dlls/ntdll/unwind.c
@@ -2046,7 +2046,7 @@ NTSTATUS WINAPI RtlVirtualUnwind2( ULONG type, ULONG_PTR base, ULONG_PTR pc,
     {
         context->Rip = *(ULONG64 *)context->Rsp;
         context->Rsp += sizeof(ULONG64);
-        *data = NULL;
+        if (data) *data = NULL;
         *handler_ret = NULL;
         return STATUS_SUCCESS;
     }
