From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: millia ampora <198710911+dappermint@users.noreply.github.com>
Date: Thu, 13 Aug 2026 01:24:33 +0400
Subject: [PATCH] 0009-nsiproxy-if-nameindex-null-deref

---
 dlls/nsiproxy.sys/ndis.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/dlls/nsiproxy.sys/ndis.c b/dlls/nsiproxy.sys/ndis.c
index c7d1f73f..94f5a0fc 100644
--- a/dlls/nsiproxy.sys/ndis.c
+++ b/dlls/nsiproxy.sys/ndis.c
@@ -29,6 +29,7 @@
 #include <sys/socket.h>
 #include <sys/ioctl.h>
 #include <unistd.h>
+#include <errno.h>
 
 #ifdef HAVE_NET_IF_H
 #include <net/if.h>
@@ -309,9 +310,16 @@ static struct if_entry *add_entry( UINT index, char *name )
 
 static unsigned int update_if_table( void )
 {
-    struct if_nameindex *indices = if_nameindex(), *entry;
+    struct if_nameindex *indices, *entry;
     unsigned int append_count = 0;
 
+    indices = if_nameindex();
+    if (!indices)
+    {
+        ERR( "if_nameindex failed, errno %d.\n", errno );
+        return 0;
+    }
+
     for (entry = indices; entry->if_index; entry++)
     {
         if (!find_entry_from_index( entry->if_index ) && add_entry( entry->if_index, entry->if_name ))
