From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: millia ampora <198710911+dappermint@users.noreply.github.com>
Date: Thu, 13 Aug 2026 01:24:33 +0400
Subject: [PATCH] 0011-winecoreaudio-division-by-zero

---
 dlls/winecoreaudio.drv/coreaudio.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/dlls/winecoreaudio.drv/coreaudio.c b/dlls/winecoreaudio.drv/coreaudio.c
index 3199887b..c2bb7650 100644
--- a/dlls/winecoreaudio.drv/coreaudio.c
+++ b/dlls/winecoreaudio.drv/coreaudio.c
@@ -744,6 +744,13 @@ static NTSTATUS unix_create_stream(void *args)
 
     stream->period = params->period;
     stream->period_frames = muldiv(params->period, stream->fmt->nSamplesPerSec, 10000000);
+
+    if (stream->period_frames == 0)
+    {
+        params->result = E_INVALIDARG;
+        goto end;
+    }
+
     stream->dev_id = dev_id_from_device(params->device);
     stream->flow = params->flow;
     stream->flags = params->flags;
